File: coding-interview/java/P13Design.java
import java.util.*;

/** 面试手撕代码（13）：设计题与数据结构实现。运行：java -ea P13Design */
public class P13Design {

    /** LC 146. 哈希 + 双向链表（哨兵）。 */
    static class LRUCache {
        private static class Node {
            int key, val;
            Node prev, next;

            Node(int k, int v) {
                key = k;
                val = v;
            }
        }

        private final int cap;
        private final Map<Integer, Node> map = new HashMap<>();
        private final Node head = new Node(0, 0), tail = new Node(0, 0);

        LRUCache(int capacity) {
            cap = capacity;
            head.next = tail;
            tail.prev = head;
        }

        private void remove(Node n) {
            n.prev.next = n.next;
            n.next.prev = n.prev;
        }

        private void addFront(Node n) {
            n.next = head.next;
            n.prev = head;
            head.next.prev = n;
            head.next = n;
        }

        int get(int key) {
            Node n = map.get(key);
            if (n == null) return -1;
            remove(n);
            addFront(n);
            return n.val;
        }

        void put(int key, int value) {
            Node n = map.get(key);
            if (n != null) {
                n.val = value;
                remove(n);
                addFront(n);
                return;
            }
            if (map.size() == cap) {
                Node lru = tail.prev;
                remove(lru);
                map.remove(lru.key);
            }
            n = new Node(key, value);
            map.put(key, n);
            addFront(n);
        }
    }

    /** LC 146 的 LinkedHashMap 版：accessOrder=true + removeEldestEntry。 */
    static class LRUCacheLinked extends LinkedHashMap<Integer, Integer> {
        private static final long serialVersionUID = 1L;
        private final int cap;

        LRUCacheLinked(int capacity) {
            super(16, 0.75f, true);
            cap = capacity;
        }

        int get(int key) {
            return super.getOrDefault(key, -1);
        }

        void put2(int key, int value) {
            super.put(key, value);
        }

        @Override
        protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
            return size() > cap;
        }
    }

    /** LC 460. freq -> LinkedHashSet（插入序 = LRU 序），minFreq 追踪。 */
    static class LFUCache {
        private final int cap;
        private final Map<Integer, Integer> kv = new HashMap<>(), kf = new HashMap<>();
        private final Map<Integer, LinkedHashSet<Integer>> buckets = new HashMap<>();
        private int minFreq = 0;

        LFUCache(int capacity) {
            cap = capacity;
        }

        private void touch(int key) {
            int f = kf.get(key);
            buckets.get(f).remove(key);
            if (buckets.get(f).isEmpty()) {
                buckets.remove(f);
                if (minFreq == f) minFreq = f + 1;
            }
            kf.put(key, f + 1);
            buckets.computeIfAbsent(f + 1, k -> new LinkedHashSet<>()).add(key);
        }

        int get(int key) {
            if (!kv.containsKey(key)) return -1;
            touch(key);
            return kv.get(key);
        }

        void put(int key, int value) {
            if (cap == 0) return;
            if (kv.containsKey(key)) {
                kv.put(key, value);
                touch(key);
                return;
            }
            if (kv.size() == cap) {
                LinkedHashSet<Integer> b = buckets.get(minFreq);
                int evict = b.iterator().next();
                b.remove(evict);
                if (b.isEmpty()) buckets.remove(minFreq);
                kv.remove(evict);
                kf.remove(evict);
            }
            kv.put(key, value);
            kf.put(key, 1);
            buckets.computeIfAbsent(1, k -> new LinkedHashSet<>()).add(key);
            minFreq = 1;
        }
    }

    /** LC 208. 26 叉数组版。 */
    static class Trie {
        private static class Node {
            Node[] ch = new Node[26];
            boolean end;
        }

        private final Node root = new Node();

        void insert(String w) {
            Node n = root;
            for (char c : w.toCharArray()) {
                int i = c - 'a';
                if (n.ch[i] == null) n.ch[i] = new Node();
                n = n.ch[i];
            }
            n.end = true;
        }

        private Node walk(String p) {
            Node n = root;
            for (char c : p.toCharArray()) {
                n = n.ch[c - 'a'];
                if (n == null) return null;
            }
            return n;
        }

        boolean search(String w) {
            Node n = walk(w);
            return n != null && n.end;
        }

        boolean startsWith(String p) {
            return walk(p) != null;
        }
    }

    /** LC 212. Trie + 网格 DFS，命中后置 end=false 去重。 */
    static List<String> findWords(char[][] board, String[] words) {
        TrieNode root = new TrieNode();
        for (String w : words) {
            TrieNode n = root;
            for (char c : w.toCharArray()) {
                if (n.ch[c - 'a'] == null) n.ch[c - 'a'] = new TrieNode();
                n = n.ch[c - 'a'];
            }
            n.word = w;
        }
        List<String> out = new ArrayList<>();
        for (int i = 0; i < board.length; i++)
            for (int j = 0; j < board[0].length; j++) dfs(board, i, j, root, out);
        return out;
    }

    private static class TrieNode {
        TrieNode[] ch = new TrieNode[26];
        String word;
    }

    private static void dfs(char[][] b, int i, int j, TrieNode parent, List<String> out) {
        if (i < 0 || j < 0 || i >= b.length || j >= b[0].length || b[i][j] == '#') return;
        char c = b[i][j];
        TrieNode n = parent.ch[c - 'a'];
        if (n == null) return;
        if (n.word != null) {
            out.add(n.word);
            n.word = null;
        }
        b[i][j] = '#';
        dfs(b, i + 1, j, n, out);
        dfs(b, i - 1, j, n, out);
        dfs(b, i, j + 1, n, out);
        dfs(b, i, j - 1, n, out);
        b[i][j] = c;
    }

    /** LC 380. 数组 + 下标哈希。 */
    static class RandomizedSet {
        private final List<Integer> vals = new ArrayList<>();
        private final Map<Integer, Integer> pos = new HashMap<>();
        private final Random rnd = new Random();

        boolean insert(int v) {
            if (pos.containsKey(v)) return false;
            pos.put(v, vals.size());
            vals.add(v);
            return true;
        }

        boolean remove(int v) {
            Integer i = pos.get(v);
            if (i == null) return false;
            int last = vals.get(vals.size() - 1);
            vals.set(i, last);
            pos.put(last, i);
            vals.remove(vals.size() - 1);
            pos.remove(v);
            return true;
        }

        int getRandom() {
            return vals.get(rnd.nextInt(vals.size()));
        }
    }

    /** LC 307. 树状数组。 */
    static class NumArray {
        private final int n;
        private final int[] nums, tree;

        NumArray(int[] a) {
            n = a.length;
            nums = a.clone();
            tree = new int[n + 1];
            for (int i = 0; i < n; i++) add(i + 1, a[i]);
        }

        private void add(int i, int d) {
            for (; i <= n; i += i & -i) tree[i] += d;
        }

        private int prefix(int i) {
            int s = 0;
            for (; i > 0; i -= i & -i) s += tree[i];
            return s;
        }

        void update(int index, int val) {
            add(index + 1, val - nums[index]);
            nums[index] = val;
        }

        int sumRange(int l, int r) {
            return prefix(r + 1) - prefix(l);
        }
    }

    /** LC 232. */
    static class MyQueue {
        private final Deque<Integer> in = new ArrayDeque<>(), out = new ArrayDeque<>();

        void push(int x) {
            in.push(x);
        }

        private void shift() {
            if (out.isEmpty()) while (!in.isEmpty()) out.push(in.pop());
        }

        int pop() {
            shift();
            return out.pop();
        }

        int peek() {
            shift();
            return out.peek();
        }

        boolean empty() {
            return in.isEmpty() && out.isEmpty();
        }
    }

    /** LC 981. 每个 key 一棵 TreeMap，floorEntry 即"<= timestamp 的最大"。 */
    static class TimeMap {
        private final Map<String, TreeMap<Integer, String>> store = new HashMap<>();

        void set(String k, String v, int t) {
            store.computeIfAbsent(k, x -> new TreeMap<>()).put(t, v);
        }

        String get(String k, int t) {
            TreeMap<Integer, String> m = store.get(k);
            if (m == null) return "";
            Map.Entry<Integer, String> e = m.floorEntry(t);
            return e == null ? "" : e.getValue();
        }
    }

    public static void main(String[] args) {
        LRUCache c = new LRUCache(2);
        c.put(1, 1);
        c.put(2, 2);
        assert c.get(1) == 1;
        c.put(3, 3);
        assert c.get(2) == -1;
        c.put(4, 4);
        assert c.get(1) == -1 && c.get(3) == 3 && c.get(4) == 4;
        LRUCacheLinked l = new LRUCacheLinked(2);
        l.put2(1, 1);
        l.put2(2, 2);
        l.get(1);
        l.put2(3, 3);
        assert l.get(2) == -1 && l.get(1) == 1;
        LFUCache f = new LFUCache(2);
        f.put(1, 1);
        f.put(2, 2);
        assert f.get(1) == 1;
        f.put(3, 3);
        assert f.get(2) == -1 && f.get(3) == 3;
        f.put(4, 4);
        assert f.get(1) == -1 && f.get(3) == 3 && f.get(4) == 4;
        Trie t = new Trie();
        t.insert("apple");
        assert t.search("apple") && !t.search("app") && t.startsWith("app");
        char[][] board = {
            "oaan".toCharArray(), "etae".toCharArray(), "ihkr".toCharArray(), "iflv".toCharArray()
        };
        List<String> found = findWords(board, new String[] {"oath", "pea", "eat", "rain"});
        Collections.sort(found);
        assert found.equals(List.of("eat", "oath"));
        RandomizedSet rs = new RandomizedSet();
        assert rs.insert(1) && !rs.remove(2) && rs.insert(2) && rs.remove(1) && rs.getRandom() == 2;
        NumArray na = new NumArray(new int[] {1, 3, 5});
        assert na.sumRange(0, 2) == 9;
        na.update(1, 2);
        assert na.sumRange(0, 2) == 8;
        MyQueue q = new MyQueue();
        q.push(1);
        q.push(2);
        assert q.peek() == 1 && q.pop() == 1 && !q.empty();
        TimeMap tm = new TimeMap();
        tm.set("foo", "bar", 1);
        tm.set("foo", "bar2", 4);
        assert tm.get("foo", 3).equals("bar")
                && tm.get("foo", 5).equals("bar2")
                && tm.get("foo", 0).isEmpty();
        System.out.println("P13Design OK");
    }
}

File: coding-interview/python/p13_design.py
"""面试手撕代码（13）：设计题与数据结构实现 —— 配套解法与测试。

https://arganzheng.life/coding-interview-design-problems-lru-lfu-trie.html
"""
from __future__ import annotations

import heapq
import random
import unittest
from bisect import bisect_left, bisect_right, insort
from collections import OrderedDict, defaultdict
from typing import Dict, List, Optional


# ---------- LC 146 LRU ----------

class _DNode:
    __slots__ = ("key", "val", "prev", "next")

    def __init__(self, key: int = 0, val: int = 0):
        self.key, self.val = key, val
        self.prev: Optional[_DNode] = None
        self.next: Optional[_DNode] = None


class LRUCache:
    """哈希表 + 双向链表（带头尾哨兵）：get / put 都 O(1)。最近使用的靠近 head。"""

    def __init__(self, capacity: int):
        self.cap = capacity
        self.map: Dict[int, _DNode] = {}
        self.head, self.tail = _DNode(), _DNode()
        self.head.next, self.tail.prev = self.tail, self.head

    def _remove(self, node: _DNode) -> None:
        node.prev.next, node.next.prev = node.next, node.prev

    def _add_front(self, node: _DNode) -> None:
        node.next, node.prev = self.head.next, self.head
        self.head.next.prev = node
        self.head.next = node

    def get(self, key: int) -> int:
        if key not in self.map:
            return -1
        node = self.map[key]
        self._remove(node)
        self._add_front(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.map:
            node = self.map[key]
            node.val = value
            self._remove(node)
            self._add_front(node)
            return
        if len(self.map) == self.cap:
            lru = self.tail.prev
            self._remove(lru)
            del self.map[lru.key]
        node = _DNode(key, value)
        self.map[key] = node
        self._add_front(node)


class LRUCacheOrdered:
    """同一题的 OrderedDict 版（面试里先说清手写版再提这个）。"""

    def __init__(self, capacity: int):
        self.cap = capacity
        self.od: "OrderedDict[int, int]" = OrderedDict()

    def get(self, key: int) -> int:
        if key not in self.od:
            return -1
        self.od.move_to_end(key)
        return self.od[key]

    def put(self, key: int, value: int) -> None:
        if key in self.od:
            self.od.move_to_end(key)
        self.od[key] = value
        if len(self.od) > self.cap:
            self.od.popitem(last=False)


# ---------- LC 460 LFU ----------

class LFUCache:
    """freq -> 该频次的 OrderedDict（按插入序即 LRU 序）；min_freq 追踪最小频次。全部 O(1)。"""

    def __init__(self, capacity: int):
        self.cap = capacity
        self.kv: Dict[int, int] = {}
        self.kf: Dict[int, int] = {}
        self.buckets: Dict[int, "OrderedDict[int, None]"] = defaultdict(OrderedDict)
        self.min_freq = 0

    def _touch(self, key: int) -> None:
        f = self.kf[key]
        del self.buckets[f][key]
        if not self.buckets[f]:
            del self.buckets[f]
            if self.min_freq == f:
                self.min_freq = f + 1
        self.kf[key] = f + 1
        self.buckets[f + 1][key] = None

    def get(self, key: int) -> int:
        if key not in self.kv:
            return -1
        self._touch(key)
        return self.kv[key]

    def put(self, key: int, value: int) -> None:
        if self.cap == 0:
            return
        if key in self.kv:
            self.kv[key] = value
            self._touch(key)
            return
        if len(self.kv) == self.cap:
            evict, _ = self.buckets[self.min_freq].popitem(last=False)
            if not self.buckets[self.min_freq]:
                del self.buckets[self.min_freq]
            del self.kv[evict], self.kf[evict]
        self.kv[key], self.kf[key] = value, 1
        self.buckets[1][key] = None
        self.min_freq = 1


# ---------- LC 208 / 212 Trie ----------

class TrieNode:
    __slots__ = ("children", "end")

    def __init__(self):
        self.children: Dict[str, TrieNode] = {}
        self.end = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            node = node.children.setdefault(ch, TrieNode())
        node.end = True

    def _walk(self, prefix: str) -> Optional[TrieNode]:
        node = self.root
        for ch in prefix:
            node = node.children.get(ch)
            if node is None:
                return None
        return node

    def search(self, word: str) -> bool:
        node = self._walk(word)
        return node is not None and node.end

    def starts_with(self, prefix: str) -> bool:
        return self._walk(prefix) is not None


def find_words(board: List[List[str]], words: List[str]) -> List[str]:
    """LC 212. 所有单词建 Trie，网格 DFS 沿 Trie 走；命中后置 end=False 去重，叶子回收剪枝。"""
    root = TrieNode()
    for w in words:
        node = root
        for ch in w:
            node = node.children.setdefault(ch, TrieNode())
        node.end = True
    m, n = len(board), len(board[0])
    out: List[str] = []

    def dfs(i, j, parent, path):
        ch = board[i][j]
        node = parent.children.get(ch)
        if node is None:
            return
        path += ch
        if node.end:
            out.append(path)
            node.end = False
        board[i][j] = "#"
        for di, dj in ((1, 0), (-1, 0), (0, 1), (0, -1)):
            x, y = i + di, j + dj
            if 0 <= x < m and 0 <= y < n and board[x][y] != "#":
                dfs(x, y, node, path)
        board[i][j] = ch
        if not node.children and not node.end:      # 叶子已用完，从 Trie 里摘掉
            del parent.children[ch]

    for i in range(m):
        for j in range(n):
            dfs(i, j, root, "")
    return out


# ---------- LC 380 O(1) 插入删除随机 ----------

class RandomizedSet:
    """数组 + 值到下标的哈希：删除时把末尾换到被删位置。"""

    def __init__(self):
        self.vals: List[int] = []
        self.pos: Dict[int, int] = {}

    def insert(self, val: int) -> bool:
        if val in self.pos:
            return False
        self.pos[val] = len(self.vals)
        self.vals.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.pos:
            return False
        i, last = self.pos[val], self.vals[-1]
        self.vals[i], self.pos[last] = last, i
        self.vals.pop()
        del self.pos[val]
        return True

    def get_random(self) -> int:
        return random.choice(self.vals)


# ---------- LC 355 设计推特 ----------

class Twitter:
    """每人一条按时间倒序的推文列表；getNewsFeed 用堆做 k 路归并取 10 条。"""

    def __init__(self):
        self.time = 0
        self.tweets: Dict[int, List[tuple]] = defaultdict(list)     # user -> [(-time, tweetId)]
        self.follows: Dict[int, set] = defaultdict(set)

    def post_tweet(self, user: int, tweet: int) -> None:
        self.tweets[user].append((-self.time, tweet))
        self.time += 1

    def get_news_feed(self, user: int) -> List[int]:
        heap = []
        for u in self.follows[user] | {user}:
            if self.tweets[u]:
                t, tid = self.tweets[u][-1]
                heap.append((t, tid, u, len(self.tweets[u]) - 1))
        heapq.heapify(heap)
        out: List[int] = []
        while heap and len(out) < 10:
            t, tid, u, idx = heapq.heappop(heap)
            out.append(tid)
            if idx > 0:
                nt, ntid = self.tweets[u][idx - 1]
                heapq.heappush(heap, (nt, ntid, u, idx - 1))
        return out

    def follow(self, a: int, b: int) -> None:
        if a != b:
            self.follows[a].add(b)

    def unfollow(self, a: int, b: int) -> None:
        self.follows[a].discard(b)


# ---------- LC 307 树状数组 ----------

class NumArray:
    """Fenwick tree：单点更新与前缀和都 O(log n)。"""

    def __init__(self, nums: List[int]):
        self.n = len(nums)
        self.nums = nums[:]
        self.tree = [0] * (self.n + 1)
        for i, x in enumerate(nums):
            self._add(i + 1, x)

    def _add(self, i: int, delta: int) -> None:
        while i <= self.n:
            self.tree[i] += delta
            i += i & -i

    def _prefix(self, i: int) -> int:
        s = 0
        while i > 0:
            s += self.tree[i]
            i -= i & -i
        return s

    def update(self, index: int, val: int) -> None:
        self._add(index + 1, val - self.nums[index])
        self.nums[index] = val

    def sum_range(self, left: int, right: int) -> int:
        return self._prefix(right + 1) - self._prefix(left)


# ---------- LC 232 两个栈实现队列 ----------

class MyQueue:
    """入栈 in、出栈 out；out 空时把 in 全倒过去。均摊 O(1)。"""

    def __init__(self):
        self.inbox: List[int] = []
        self.outbox: List[int] = []

    def push(self, x: int) -> None:
        self.inbox.append(x)

    def _shift(self) -> None:
        if not self.outbox:
            while self.inbox:
                self.outbox.append(self.inbox.pop())

    def pop(self) -> int:
        self._shift()
        return self.outbox.pop()

    def peek(self) -> int:
        self._shift()
        return self.outbox[-1]

    def empty(self) -> bool:
        return not self.inbox and not self.outbox


# ---------- LC 981 基于时间的键值存储 ----------

class TimeMap:
    """每个 key 一个按时间递增的列表，get 用 bisect_right - 1。"""

    def __init__(self):
        self.store: Dict[str, List[tuple]] = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.store[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        arr = self.store.get(key, [])
        i = bisect_right(arr, (timestamp, chr(0x10FFFF)))
        return arr[i - 1][1] if i else ""


class SortedList:
    """有序表的极简替身（bisect + insort，插入 O(n)）——面试里说明 Java 用 TreeMap、Python 用 sortedcontainers。"""

    def __init__(self):
        self.a: List[int] = []

    def add(self, x: int) -> None:
        insort(self.a, x)

    def remove(self, x: int) -> None:
        i = bisect_left(self.a, x)
        del self.a[i]

    def rank(self, x: int) -> int:
        return bisect_left(self.a, x)


class Tests(unittest.TestCase):
    def test_lru(self):
        for cls in (LRUCache, LRUCacheOrdered):
            c = cls(2)
            c.put(1, 1); c.put(2, 2)
            self.assertEqual(c.get(1), 1)
            c.put(3, 3)
            self.assertEqual(c.get(2), -1)
            c.put(4, 4)
            self.assertEqual(c.get(1), -1)
            self.assertEqual(c.get(3), 3)
            self.assertEqual(c.get(4), 4)

    def test_lfu(self):
        c = LFUCache(2)
        c.put(1, 1); c.put(2, 2)
        self.assertEqual(c.get(1), 1)
        c.put(3, 3)                       # 淘汰 2（频次 1，最少）
        self.assertEqual(c.get(2), -1)
        self.assertEqual(c.get(3), 3)
        c.put(4, 4)                       # 1 与 3 频次都是 2，淘汰更久没用的 1
        self.assertEqual(c.get(1), -1)
        self.assertEqual(c.get(3), 3)
        self.assertEqual(c.get(4), 4)

    def test_trie(self):
        t = Trie()
        t.insert("apple")
        self.assertTrue(t.search("apple"))
        self.assertFalse(t.search("app"))
        self.assertTrue(t.starts_with("app"))
        board = [list("oaan"), list("etae"), list("ihkr"), list("iflv")]
        self.assertEqual(sorted(find_words(board, ["oath", "pea", "eat", "rain"])), ["eat", "oath"])
        self.assertEqual(find_words([list("ab"), list("cd")], ["abcb"]), [])

    def test_randomized_set(self):
        s = RandomizedSet()
        self.assertTrue(s.insert(1)); self.assertFalse(s.remove(2)); self.assertTrue(s.insert(2))
        self.assertIn(s.get_random(), (1, 2))
        self.assertTrue(s.remove(1)); self.assertFalse(s.insert(2))
        self.assertEqual(s.get_random(), 2)

    def test_twitter(self):
        tw = Twitter()
        tw.post_tweet(1, 5)
        self.assertEqual(tw.get_news_feed(1), [5])
        tw.follow(1, 2); tw.post_tweet(2, 6)
        self.assertEqual(tw.get_news_feed(1), [6, 5])
        tw.unfollow(1, 2)
        self.assertEqual(tw.get_news_feed(1), [5])

    def test_fenwick(self):
        na = NumArray([1, 3, 5])
        self.assertEqual(na.sum_range(0, 2), 9)
        na.update(1, 2)
        self.assertEqual(na.sum_range(0, 2), 8)
        self.assertEqual(na.sum_range(1, 1), 2)

    def test_queue_timemap(self):
        q = MyQueue()
        q.push(1); q.push(2)
        self.assertEqual(q.peek(), 1)
        self.assertEqual(q.pop(), 1)
        self.assertFalse(q.empty())
        tm = TimeMap()
        tm.set("foo", "bar", 1)
        self.assertEqual(tm.get("foo", 1), "bar")
        self.assertEqual(tm.get("foo", 3), "bar")
        tm.set("foo", "bar2", 4)
        self.assertEqual(tm.get("foo", 4), "bar2")
        self.assertEqual(tm.get("foo", 5), "bar2")
        self.assertEqual(tm.get("foo", 0), "")
        sl = SortedList()
        for x in (5, 1, 3):
            sl.add(x)
        self.assertEqual(sl.a, [1, 3, 5]); sl.remove(3); self.assertEqual(sl.rank(5), 1)


if __name__ == "__main__":
    unittest.main(verbosity=1)
